diff --git a/lib/parse.js b/lib/parse.js
index 3a6988e629f52aee6609dc5ec290a5e9ff47cecb..9a04fad46b66465e11b7efaa9bd35742ceb9f866 100644
--- a/lib/parse.js
+++ b/lib/parse.js
@@ -143,6 +143,9 @@ const parse = (input, options = {}) => {
      */
 
     if (value === CHAR_LEFT_PARENTHESES) {
+      if (stack.length >= 101) {
+        throw new SyntaxError('Nesting exceeds maximum depth 100');
+      }
       block = push({ type: 'paren', nodes: [] });
       stack.push(block);
       push({ type: 'text', value });
@@ -195,6 +198,9 @@ const parse = (input, options = {}) => {
      */
 
     if (value === CHAR_LEFT_CURLY_BRACE) {
+      if (stack.length >= 101) {
+        throw new SyntaxError('Nesting exceeds maximum depth 100');
+      }
       depth++;
 
       const dollar = prev.value && prev.value.slice(-1) === '$' || block.dollar === true;
